frontend/server/src/Controllers/Course.php

```diff
@@ -561,10 +561,43 @@ public static function apiClone(\OmegaUp\Request $r): array {
             fn (string $name) => \OmegaUp\Validators::stringNonEmpty($name)
         );
         $token = $r->ensureOptionalString('token');
+        $originalCourse = self::validateCourseExists($courseAlias);
+        $decodedToken = null;
         if (!is_null($token)) {
-            // TODO: Add function to decode token, and all the validations
+            try {
+                $decodedToken = \OmegaUp\SecurityTools::getDecodedCloneCourseToken(
+                    $token,
+                    $courseAlias
+                );
+            } catch (\OmegaUp\Exceptions\TokenDecodeException $e) {
+                $result = 'token_corrupted';
+                if (isset($e->claims)) {
+                    $result = 'token_expired';
+                }
+                self::$log->error(
+                    "Error decoding token for course {$courseAlias}",
+                    $e
```

Comment:
```suggestion
                    "Error validating token for course {$courseAlias}: $e"
```

si no, no sale en el log. el log lo puedes ver en `frontend/tests/runfiles/shard-0/omegaup.log`. puedes buscar la cadena `===== Start CourseCloneTest::testDecodeCloneCourseToken` para ver por qué falló.

Reply:
Listo